packages/schema/src/idl.rs

```diff
@@ -0,0 +1,104 @@
+//! The Cosmwasm IDL (Interface Description Language)
+
+use std::collections::BTreeMap;
+
+use schemars::schema::RootSchema;
+use thiserror::Error;
+
+/// The version of the CosmWasm IDL.
+///
+/// Follows Semantic Versioning 2.0.0: https://semver.org/
+///
+/// To determine if a change is breaking, assume consumers allow unknown fields.
+pub const IDL_VERSION: &str = "1.0.0";
+
+/// Rust representation of a contract's API.
+pub struct Api {
+    pub contract_name: String,
+    pub contract_version: String,
+    pub instantiate: RootSchema,
+    pub execute: RootSchema,
+    pub query: RootSchema,
+    pub migrate: Option<RootSchema>,
+    pub sudo: Option<RootSchema>,
+    /// A mapping of query variants to response types
+    pub responses: BTreeMap<String, RootSchema>,
+}
+
+impl Api {
+    pub fn render(self) -> JsonApi<'static> {
+        let mut json_api = JsonApi {
+            contract_name: self.contract_name,
+            contract_version: self.contract_version,
+            idl_version: IDL_VERSION,
+            instantiate: self.instantiate,
+            execute: self.execute,
+            query: self.query,
+            migrate: self.migrate,
+            sudo: self.sudo,
+            responses: self.responses,
+        };
+
+        if let Some(metadata) = &mut json_api.instantiate.schema.metadata {
+            metadata.title = Some("InstantiateMsg".to_string());
+        }
+        if let Some(metadata) = &mut json_api.execute.schema.metadata {
+            metadata.title = Some("ExecuteMsg".to_string());
+        }
+        if let Some(metadata) = &mut json_api.query.schema.metadata {
+            metadata.title = Some("QueryMsg".to_string());
+        }
+        if let Some(migrate) = &mut json_api.migrate {
+            if let Some(metadata) = &mut migrate.schema.metadata {
+                metadata.title = Some("MigrateMsg".to_string());
+            }
+        }
+        if let Some(sudo) = &mut json_api.sudo {
+            if let Some(metadata) = &mut sudo.schema.metadata {
+                metadata.title = Some("SudoMsg".to_string());
+            }
+        }
+
+        json_api
+    }
+}
+
+/// A JSON representation of a contract's API.
+#[derive(serde::Serialize)]
+pub struct JsonApi<'v> {
```

Comment:
Do we really need this lifetime? Can't we just use owned data for simplicity?

Reply:
Let's make it owned. Structs with lifetimes are a bit annoying to work with as this easily propagates to a lot of places. I'd only do this if it brings serious performance benefits.